driver/src/main/java/org/neo4j/driver/v1/TransactionConfig.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2002-2018 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.v1;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Objects;
+
+import org.neo4j.driver.internal.util.Extract;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+import static org.neo4j.driver.internal.util.Preconditions.checkArgument;
+
+/**
+ * Configuration object containing settings for explicit and auto-commit transactions.
+ * Instances are immutable and can be reused for multiple transactions.
+ * <p>
+ * Configuration is supported for:
+ * <ul>
+ * <li>queries executed in auto-commit transactions - using various overloads of {@link Session#run(String, TransactionConfig)} and
+ * {@link Session#runAsync(String, TransactionConfig)}</li>
+ * <li>transactions started by transaction functions - using {@link Session#readTransaction(TransactionWork, TransactionConfig)},
+ * {@link Session#writeTransaction(TransactionWork, TransactionConfig)}, {@link Session#readTransactionAsync(TransactionWork, TransactionConfig)} and
+ * {@link Session#writeTransactionAsync(TransactionWork, TransactionConfig)}</li>
+ * <li>explicit transactions - using {@link Session#beginTransaction(TransactionConfig)} and {@link Session#beginTransactionAsync(TransactionConfig)}</li>
+ * </ul>
+ * <p>
+ * Creation of configuration objects can be done using the builder API:
+ * <pre>
+ * {@code
+ * Map<String, Object> metadata = new HashMap<>();
+ * metadata.put("type", "update user");
+ * metadata.put("application", "my application");
+ *
+ * TransactionConfig config = TransactionConfig.builder()
+ *                 .withTimeout(Duration.ofSeconds(4))
+ *                 .withMetadata(metadata)
+ *                 .build();
+ * }
+ * </pre>
+ *
+ * @see Session
+ */
+public class TransactionConfig
+{
+    private static final TransactionConfig EMPTY = builder().build();
```

Comment:
Since we now have this builder style, shall we also give Driver config a builder style?
So we deprecate the old method and then provides a new builder API. WDYT?

Reply:
I would prefer to just change it in the next major release, tbh